File: app/src/main/java/com/shrikantbadwaik/tmdb/view/moviedetails/MovieDetailsActivity.java
package com.shrikantbadwaik.tmdb.view.moviedetails;

import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;
import android.support.annotation.Nullable;

import com.shrikantbadwaik.tmdb.BR;
import com.shrikantbadwaik.tmdb.R;
import com.shrikantbadwaik.tmdb.data.model.Image;
import com.shrikantbadwaik.tmdb.databinding.MovieDetailsActivityBinding;
import com.shrikantbadwaik.tmdb.view.base.BaseActivity;
import com.shrikantbadwaik.tmdb.viewmodel.MovieDetailsViewModel;

import java.util.List;

import javax.inject.Inject;

import dagger.android.AndroidInjection;

public class MovieDetailsActivity extends BaseActivity<MovieDetailsActivityBinding, MovieDetailsViewModel>
        implements MovieDetailsView {
    @Inject
    ViewModelProvider.Factory viewModelFactory;
    @Inject
    MovieDetailsViewPagerAdapter adapter;

    private MovieDetailsViewModel viewModel;
    private MovieDetailsActivityBinding activityBinding;

    @Override
    protected void injectDependency() {
        AndroidInjection.inject(this);
    }

    @Override
    protected int layoutResource() {
        return R.layout.movie_details_activity;
    }

    @Override
    protected MovieDetailsViewModel viewModel() {
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(MovieDetailsViewModel.class);
        return viewModel;
    }

    @Override
    protected int bindingVariable() {
        return BR.viewModel;
    }

    @Override
    protected void initView() {
        activityBinding = viewDataBinding();
        viewModel.attachView(this);

        setupToolbar();
        activityBinding.movieDetailsActivityViewPager.setAdapter(adapter);
        viewModel.intentExtras(getIntent().getExtras());
        subscribeToLiveData();
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }

    private void setupToolbar() {
        setSupportActionBar(activityBinding.movieDetailsActivityToolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setDisplayShowHomeEnabled(true);
            getSupportActionBar().setDisplayShowTitleEnabled(false);
        }
    }

    private void subscribeToLiveData() {
        viewModel.getImageListLiveData().observe(this, new Observer<List<Image>>() {
            @Override
            public void onChanged(@Nullable List<Image> imageList) {
                viewModel.addImagesToObservable(imageList);
            }
        });
    }
}

File: app/src/main/java/com/shrikantbadwaik/tmdb/view/moviedetails/MovieDetailsActivity.kt
package com.shrikantbadwaik.tmdb.view.moviedetails

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProvider
import android.arch.lifecycle.ViewModelProviders
import com.shrikantbadwaik.tmdb.BR
import com.shrikantbadwaik.tmdb.R
import com.shrikantbadwaik.tmdb.databinding.ActivityMovieDetailsBinding
import com.shrikantbadwaik.tmdb.view.base.BaseActivity
import com.shrikantbadwaik.tmdb.viewmodel.MovieDetailsActivityViewModel
import dagger.android.AndroidInjection
import javax.inject.Inject

class MovieDetailsActivity : BaseActivity<ActivityMovieDetailsBinding, MovieDetailsActivityViewModel>(),
    MovieDetailsView {
    @Inject
    lateinit var factory: ViewModelProvider.Factory
    @Inject
    lateinit var adapter: MovieDetailsPagerAdapter

    private lateinit var activityBinding: ActivityMovieDetailsBinding
    private lateinit var movieDetailsActivityViewModel: MovieDetailsActivityViewModel

    override fun injectDependency() {
        AndroidInjection.inject(this)
    }

    override val layoutResource: Int = R.layout.activity_movie_details

    override val viewModel: MovieDetailsActivityViewModel
        get() {
            movieDetailsActivityViewModel =
                    ViewModelProviders.of(this, factory).get(MovieDetailsActivityViewModel::class.java)
            return movieDetailsActivityViewModel
        }

    override val variableId: Int = BR.viewModel

    override fun initView() {
        activityBinding = viewDataBinding()
        movieDetailsActivityViewModel.attachView(this)
        setupToolbar()
        activityBinding.activityMovieDetailsViewPager.adapter = adapter
        movieDetailsActivityViewModel.intentExtras(intent.extras)
        movieDetailsActivityViewModel.getImageListLiveData()
            .observe(this, Observer { movieDetailsActivityViewModel.addImagesToObservable(it) })
    }

    override fun onSupportNavigateUp(): Boolean {
        onBackPressed()
        return true
    }

    private fun setupToolbar() {
        setSupportActionBar(activityBinding.activityMovieDetailsToolbar)
        supportActionBar?.let { actionBar ->
            actionBar.setDisplayHomeAsUpEnabled(true)
            actionBar.setDisplayShowHomeEnabled(true)
            actionBar.setDisplayShowTitleEnabled(false)
        }
    }
}